keps/sig-api-machinery/3962-mutating-admission-policies/README.md

```diff
@@ -0,0 +1,1348 @@
+- option 2: the lexicographical ordering of the resource names
+  - Pros:
+    - Align with the behavior with mutating webhook
+    - User has a hacky way to enforce the order if it is really needed
+  - Cons:
+    - User has to be mindful on the order if there is dependency existing
+- option 3: ordering based on input from user
+  - Pros:
+    - User has more control over ordering
+  - Cons:
+    - Put more responsibility on user side
+
+Considering it would be easier to go with random order and then switch to a particular order, 
+the random order could be applied in phase 1(first alpha) and provide more time on discussion if further opinion raised.
+<<open to discuss>>
+
+#### Metrics
+Goals:
+- Parity with validating admission policy metrics 
+  - Should include counter of deny, success violations 
+  - Label by {policy, policy binding, mutation expression} identifiers
+- Counters for number of policy definitions and policy bindings in cluster 
+  - Label by state (active vs. error), enforcement action (deny, warn)
+- Counters for Variable Composition 
+  - Should include a counter of variable resolutions to measure time saved. 
+  - Label by policy identifier
+
+
+### Phase 2
+
+All these capabilities are required and should be discussed thoughtfully before Beta, but will not be implemented in the first alpha release of this enhancement due to the size and complexity of this enhancement.
+
+#### Construct Type Enforcement
+
+The type alias Object and its descendants, in Phase 2, are now real types that derive from the resolved OpenAPI schemas. 
+If any type violations happen in the constructed object, the CEL checker will raise the errors before the expressions evaluate.
+
+For bigger schemas, the construction of CEL types can be expensive. It is recorded to take ~100ms to resolve and parse apps/v1.Deployment. 
+Optimizations like caching or lazy sub-schema resolution can be candidates of beta/GA graduation criteria.
+
+#### Unsetting values
+
+Since there is no field manager used for the merge, the server side apply merge
+algorithm will only add and replace values. This is because server side apply is
+designed to only unset values that were previously owned by the field manager
+but excluded from the apply configuration.
+
+To work around this limitation, we will take advantage of CEL's optional type
+feature to make it possible to express that a value should be unset. For example:
+
+```cel
+Object{
+  spec: Object.spec{
+    ?fieldToRemoveIfPresent: optional.none()
+  }
+}
+```
+The policy declaratively describes that, in the final object, if fieldToRemoveIfPresent presents it should be removed or no-op otherwise.
+
+The optional.none() function creates a CEL object defined as optional(T) where T is the type of the receiving field. 
+In conjunction with standard CEL macros, simple expressions can perform more complicated and precise operations.
+
+For example, to remove the env of a sidecar container, filter by its name.
```

Comment:
how do I remove an entire container with an envvar named `remove-this-container`?